File: csharp/Resources/AddonsResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Environment add-on Wave 3 operations
    /// </summary>
    public class AddonsResource
    {
        private readonly AuthdogClient _client;

        public AddonsResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> ListAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/addons");

        public JObject List(string tenantId, string environmentId) =>
            ListAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> SaveAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/addons",
                body);

        public JObject Save(string tenantId, string environmentId, object body) =>
            SaveAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteAsync(string tenantId, string environmentId, string provider) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"{AuthdogClient.Env(tenantId, environmentId)}/addons/{provider}");

        public JObject Delete(string tenantId, string environmentId, string provider) =>
            DeleteAsync(tenantId, environmentId, provider).GetAwaiter().GetResult();
    }
}

File: java/src/main/java/com/authdog/resources/AddonsResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Environment addon operations.
 */
public final class AddonsResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create an addons helper.
     * @param clientParam Authdog client
     */
    public AddonsResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * List addons in an environment.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return addons envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode list(final String tenantId,
                         final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId) + "/addons",
                null, null, JsonNode.class);
    }

    /**
     * Create or update an addon.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return addon envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode save(final String tenantId,
                         final String environmentId,
                         final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId) + "/addons",
                body, null, JsonNode.class);
    }

    /**
     * Delete an addon.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param provider addon provider
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode delete(final String tenantId,
                           final String environmentId,
                           final String provider)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/addons/" + provider,
                null, null, JsonNode.class);
    }
}
